File: apps/web/src/lib/store/variant-options.ts
import type { ProductOptionGroup, ProductVariant } from "@/lib/api";

export type VariantSelection = Record<string, string>;

const SIZE_KEYS = new Set(["size", "размер", "размер обуви", "размер ремня", "shoe size", "belt size"]);
const COLOR_KEYS = new Set(["color", "цвет", "камуфляж", "camouflage", "расцветка"]);
const WAIST_KEYS = new Set(["waist", "талия", "обхват талии"]);

function extractSizeFromName(name?: string | null): string | null {
  const match = name?.trim().match(/\(([^)]+)\)\s*$/);
  if (!match) {
    return null;
  }
  const candidate = match[1].trim().replace(/–/g, "-").replace(/\s*-\s*/g, "-");
  return /^(?:\d{2,3}(?:-\d{2,3})?|[XxSsMmLl]{1,3})$/.test(candidate) ? candidate : null;
}

function canonicalAttributeKey(rawKey: string): "size" | "color" | "waist" | null {
  const key = rawKey.trim().toLocaleLowerCase();
  if (!key) {
    return null;
  }
  if (SIZE_KEYS.has(key) || (key.includes("размер") && !key.includes("цвет"))) {
    return "size";
  }
  if (COLOR_KEYS.has(key) || key.includes("цвет") || key.includes("камуфляж") || key.includes("camouflage")) {
    return "color";
  }
  if (WAIST_KEYS.has(key) || key.includes("талия")) {
    return "waist";
  }
  return null;
}

function normalizeVariantAttributes(
  attributes: ProductVariant["attributes"] | null | undefined,
  name?: string | null,
): Record<string, string> {
  const normalized: Record<string, string> = {};
  for (const [key, value] of Object.entries(attributes ?? {})) {
    if (value == null) {
      continue;
    }
    normalized[key] = String(value);
  }
  for (const [key, value] of Object.entries(normalized)) {
    const canonical = canonicalAttributeKey(key);
    const trimmed = value.trim();
    if (canonical && trimmed && !normalized[canonical]?.trim()) {
      normalized[canonical] = trimmed;
    }
  }
  if (!normalized.size?.trim()) {
    const fromName = extractSizeFromName(name);
    if (fromName) {
      normalized.size = fromName;
    }
  }
  return normalized;
}

function normalizeColor(value: string): string {
  const lower = value.toLowerCase();
  const map: Record<string, string> = {
    multicam: "Multicam",
    coyote: "Coyote",
    olive: "Olive",
    black: "Black",
    "ranger green": "Ranger Green",
    woodland: "Woodland",
  };
  for (const [keyword, label] of Object.entries(map)) {
    if (lower.includes(keyword)) {
      return label;
    }
  }
  return value.trim();
}

export function variantOptionValues(variant: ProductVariant): VariantSelection {
  const selected: VariantSelection = {};
  const attrs = normalizeVariantAttributes(variant.attributes, variant.name);

  const colorRaw = attrs.color || attrs.camouflage;
  if (colorRaw?.trim()) {
    selected.color = normalizeColor(colorRaw);
  }
  if (attrs.size?.trim()) {
    selected.size = attrs.size.trim();
  }
  if (attrs.waist?.trim()) {
    selected.waist = `W${attrs.waist.trim()}`;
  }

  return selected;
}

export function usesStructuredSelector(
  optionGroups: ProductOptionGroup[],
  variantCount?: number,
): boolean {
  if (variantCount != null && variantCount <= 1) {
    return false;
  }
  if (optionGroups.length === 0) {
    return false;
  }
  const totalValues = optionGroups.reduce((sum, group) => sum + group.values.length, 0);
  return totalValues > 1;
}

export function resolveVariant(
  variants: ProductVariant[],
  selection: VariantSelection,
): ProductVariant | null {
  if (!selection || Object.keys(selection).length === 0) {
    return null;
  }

  return (
    variants.find((variant) => {
      const options = variantOptionValues(variant);
      return Object.entries(selection).every(([key, value]) => options[key] === value);
    }) ?? null
  );
}

export function pickDefaultSelection(
  variants: ProductVariant[],
  optionGroups: ProductOptionGroup[],
): VariantSelection {
  if (variants.length === 0 || optionGroups.length === 0) {
    return {};
  }

  const defaultVariant = pickDefaultVariant(variants) ?? variants[0];
  const defaultOptions = variantOptionValues(defaultVariant);
  const selection: VariantSelection = {};

  for (const group of optionGroups) {
    if (defaultOptions[group.key]) {
      selection[group.key] = defaultOptions[group.key];
    }
  }

  for (const group of optionGroups) {
    if (selection[group.key]) {
      continue;
    }
    let fallback: string | undefined;
    for (const value of group.values) {
      const trial = { ...selection, [group.key]: value };
      const match = resolveVariant(variants, trial);
      if (!match) {
        continue;
      }
      if (match.in_stock) {
        selection[group.key] = value;
        break;
      }
      fallback ??= value;
    }
    if (!selection[group.key] && fallback) {
      selection[group.key] = fallback;
    }
  }

  return selection;
}

export function getValueState(
  variants: ProductVariant[],
  selection: VariantSelection,
  axisKey: string,
  value: string,
): { exists: boolean; inStock: boolean } {
  const trial = { ...selection, [axisKey]: value };
  const matches = variants.filter((variant) => {
    const options = variantOptionValues(variant);
    return Object.entries(trial).every(([key, item]) => options[key] === item);
  });

  return {
    exists: matches.length > 0,
    inStock: matches.some((variant) => variant.in_stock),
  };
}

export function extractColorOptions(optionGroups: ProductOptionGroup[]): string[] {
  const colorGroup = optionGroups.find((group) => group.key === "color");
  return colorGroup?.values ?? [];
}

export function getVariantPriceRange(variants: ProductVariant[]): {
  min: number;
  max: number;
} | null {
  if (variants.length === 0) {
    return null;
  }
  const prices = variants.map((variant) => variant.price_cents);
  const min = Math.min(...prices);
  const max = Math.max(...prices);
  return { min, max };
}

export function pickDefaultVariant(variants: ProductVariant[]): ProductVariant | null {
  if (variants.length === 0) {
    return null;
  }
  const sorted = [...variants].sort((a, b) => a.sort_order - b.sort_order);
  const defaultVariant = sorted.find((variant) => variant.is_default) ?? sorted[0];
  return defaultVariant.in_stock
    ? defaultVariant
    : sorted.find((variant) => variant.in_stock) ?? defaultVariant;
}

export function getColorOptionsFromVariants(variants: ProductVariant[]): string[] {
  const values = new Set<string>();
  for (const variant of variants) {
    const color = variantOptionValues(variant).color;
    if (color) {
      values.add(color);
    }
  }
  return [...values];
}

const SUMMARY_LABELS: Record<string, string> = {
  size: "Размер",
  color: "Цвет",
  waist: "Талия",
};

export function formatVariantSelectionSummary(
  variant: ProductVariant,
  structured: boolean,
  selection: VariantSelection,
): string {
  const parts: string[] = [];

  if (structured) {
    for (const [key, label] of Object.entries(SUMMARY_LABELS)) {
      const value = selection[key]?.trim();
      if (value) {
        parts.push(`${label}: ${value}`);
      }
    }
  } else {
    const attrs = normalizeVariantAttributes(variant.attributes, variant.name);
    if (attrs.size?.trim()) {
      parts.push(`Размер: ${attrs.size.trim()}`);
    }
    if (attrs.color?.trim()) {
      parts.push(`Цвет: ${attrs.color.trim()}`);
    }
    if (variant.name && variant.name !== "Default" && parts.length === 0) {
      parts.push(variant.name);
    }
  }

  return parts.join(" · ");
}

File: apps/api/app/features/catalog/domain/variant_options.py
"""Structured variant option groups for storefront selectors (ADR-011)."""

from dataclasses import dataclass

from app.features.catalog.domain.entities import ProductVariant
from app.features.catalog.domain.variant_filter import _normalize_color

_OPTION_AXES: tuple[tuple[str, str, tuple[str, ...]], ...] = (
    ("color", "Цвет / камуфляж", ("color", "camouflage")),
    ("size", "Размер", ("size",)),
    ("waist", "Талия", ("waist",)),
)


@dataclass(frozen=True, slots=True)
class ProductOptionGroup:
    key: str
    label: str
    values: tuple[str, ...]


def _axis_value(variant: ProductVariant, attribute_keys: tuple[str, ...]) -> str | None:
    for key in attribute_keys:
        raw = (variant.attributes.get(key) or "").strip()
        if not raw:
            continue
        if key in {"color", "camouflage"}:
            normalized = _normalize_color(raw) or raw
            return normalized
        return raw
    return None


def variant_option_values(variant: ProductVariant) -> dict[str, str]:
    """Map axis key -> selected value for one variant."""
    selected: dict[str, str] = {}
    for axis_key, _label, attribute_keys in _OPTION_AXES:
        value = _axis_value(variant, attribute_keys)
        if value:
            selected[axis_key] = value
    return selected


def build_option_groups(variants: tuple[ProductVariant, ...]) -> tuple[ProductOptionGroup, ...]:
    if not variants:
        return ()

    groups: list[ProductOptionGroup] = []
    for axis_key, label, attribute_keys in _OPTION_AXES:
        values: list[str] = []
        seen: set[str] = set()
        for variant in sorted(variants, key=lambda item: item.sort_order):
            value = _axis_value(variant, attribute_keys)
            if value and value not in seen:
                seen.add(value)
                values.append(value)
        if values:
            groups.append(ProductOptionGroup(key=axis_key, label=label, values=tuple(values)))
    return tuple(groups)


def uses_structured_selector(
    option_groups: tuple[ProductOptionGroup, ...],
    *,
    variant_count: int | None = None,
) -> bool:
    if variant_count is not None and variant_count <= 1:
        return False
    if not option_groups:
        return False
    total_values = sum(len(group.values) for group in option_groups)
    return total_values > 1


def resolve_variant(
    variants: tuple[ProductVariant, ...],
    selected: dict[str, str],
) -> ProductVariant | None:
    if not variants or not selected:
        return None
    for variant in variants:
        options = variant_option_values(variant)
        if all(options.get(key) == value for key, value in selected.items()):
            return variant
    return None


def pick_default_selection(
    variants: tuple[ProductVariant, ...],
    option_groups: tuple[ProductOptionGroup, ...],
) -> dict[str, str]:
    if not variants or not option_groups:
        return {}

    default = next((variant for variant in variants if variant.is_default), variants[0])
    default_options = variant_option_values(default)
    selection = {group.key: default_options[group.key] for group in option_groups if group.key in default_options}

    for group in option_groups:
        if group.key in selection:
            continue
        for value in group.values:
            trial = {**selection, group.key: value}
            if resolve_variant(variants, trial) is not None:
                selection[group.key] = value
                break
    return selection
